admin can override the exists pkg

```
$ tnpm publish
请求服务器发布模块...
[WARNING] Command failed: npm http PUT http://registry.npm.taobao.net/tair
npm http 403 http://registry.npm.taobao.net/tair
npm http PUT http://registry.npm.taobao.net/tair
npm http 403 http://registry.npm.taobao.net/tair
npm ERR! publish Failed PUT response 403
npm ERR! Error: no_perms Current user can not publish this module: tair
npm ERR!     at RegClient.<anonymous> (/opt/taobao/install/node.js/lib/node_modules/npm/node_modules/npm-registry-client/lib/request.js:272:14)
npm ERR!     at Request.self.callback (/opt/taobao/install/node.js/lib/node_modules/npm/node_modules/request/request.js:129:22)
npm ERR!     at Request.EventEmitter.emit (events.js:98:17)
npm ERR!     at Request.<anonymous> (/opt/taobao/install/node.js/lib/node_modules/npm/node_modules/request/request.js:873:14)
npm ERR!     at Request.EventEmitter.emit (events.js:117:20)
npm ERR!     at IncomingMessage.<anonymous> (/opt/taobao/install/node.js/lib/node_modules/npm/node_modules/request/request.js:824:12)
npm ERR!     at IncomingMessage.EventEmitter.emit (events.js:117:20)
npm ERR!     at _stream_readable.js:920:16
npm ERR!     at process._tickCallback (node.js:415:13)
npm ERR! If you need help, you may report this log at:
npm ERR!     <http://github.com/isaacs/npm/issues>
npm ERR! or email it to:
npm ERR!     <npm-@googlegroups.com>

npm ERR! System Linux 2.6.18-274.3.1.tb443.el5
npm ERR! command "/opt/taobao/install/node.js/bin/node" "/opt/taobao/install/node.js/bin/npm" "publish" "." "--registry=http://registry.npm.taobao.net"
npm ERR! cwd /home/admin/npm-web/target/npm-web/tmp/1390457484515_suqian.yf_vw7e4tj0
npm ERR! node -v v0.10.21
npm ERR! npm -v 1.3.11
npm ERR! 
npm ERR! Additional logging details can be found in:
npm ERR!     /home/admin/npm-web/target/npm-web/tmp/1390457484515_suqian.yf_vw7e4tj0/npm-debug.log
npm ERR! not ok code 0

```

--- a/controllers/registry/module.js
+++ b/controllers/registry/module.js
@@ -316,10 +316,8 @@ exports.upload = function (req, res, next) {
     if (err || !mod) {
       return next(err);
     }
-    var match = mod.package.maintainers.filter(function (item) {
-      return item.name === username;
-    });
-    if (match.length === 0 || mod.name !== name) {
+
+    if (!common.isMaintainer(req, mod.package.maintainers) || mod.name !== name) {
       return res.json(403, {
         error: 'no_perms',
         reason: 'Current user can not publish this module'
@@ -589,14 +587,8 @@ exports.add = function (req, res, next) {
   var username = req.session.name;
   var name = req.params.name;
   var pkg = req.body || {};
-  var maintainers = pkg.maintainers || [];
-  var match = maintainers.filter(function (item) {
-    return item.name === username;
-  });
 
-  debug('add module %s maintainers match: %j, current user: %s', name, match, username);
-
-  if (match.length === 0) {
+  if (!common.isMaintainer(req, pkg.maintainers)) {
     return res.json(403, {
       error: 'no_perms',
       reason: 'Current user can not publish this module'
@@ -640,11 +632,8 @@ exports.add = function (req, res, next) {
   ep.all('latest', 'next', function (latestMod, nextMod) {
     var maintainers = latestMod && latestMod.package.maintainers.length > 0 ?
       latestMod.package.maintainers : nextMod.package.maintainers;
-    var match = maintainers.filter(function (item) {
-      return item.name === username;
-    });
 
-    if (match.length === 0) {
+    if (!common.isMaintainer(req, maintainers)) {
       return res.json(403, {
         error: 'no_perms',
         reason: 'Current user can not publish this module'
@@ -681,12 +670,9 @@ exports.removeWithVersions = function (req, res, next) {
     if (!mods || !mods.length) {
       return next();
     }
-    //TODO replace this maintainer check
-    var match = mods[0].package.maintainers.filter(function (item) {
-      return item.name === username;
-    });
 
-    if (!match.length || mods[0].name !== name) {
+    var firstMod = mods[0];
+    if (!common.isMaintainer(req, firstMod.package.maintainers) || firstMod.name !== name) {
       return res.json(403, {
         error: 'no_perms',
         reason: 'Current user can not update this module'
@@ -724,11 +710,8 @@ exports.removeTar = function (req, res, next) {
     if (!mod) {
       return next();
     }
-    //TODO replace this maintainer check
-    var match = mod.package.maintainers.filter(function (item) {
-      return item.name === username;
-    });
-    if (!match.length || mod.name !== name) {
+
+    if (!common.isMaintainer(req, mod.package.maintainers) || mod.name !== name) {
       return res.json(403, {
         error: 'no_perms',
         reason: 'Current user can not delete this tarball'
@@ -759,15 +742,8 @@ exports.removeAll = function (req, res, next) {
     if (!mod) {
       return next();
     }
-    //TODO replace this maintainer check
-    var match = mod.package.maintainers.filter(function (item) {
-      return item.name === username;
-    });
-    if (req.session.isAdmin) {
-      match.push({name: username});
-    }
 
-    if (!match.length || mod.name !== name) {
+    if (!common.isMaintainer(req, mod.package.maintainers) || mod.name !== name) {
       return res.json(403, {
         error: 'no_perms',
         reason: 'Current user can not delete this tarball'

--- a/lib/common.js
+++ b/lib/common.js
@@ -41,3 +41,17 @@ exports.setDownloadURL = function (pkg, req, host) {
 exports.isAdmin = function (username) {
   return typeof config.admins[username] === 'string';
 };
+
+exports.isMaintainer = function (req, maintainers) {
+  if (req.session.isAdmin) {
+    return true;
+  }
+
+  var username = req.session.name;
+  maintainers = maintainers || [];
+  var match = maintainers.filter(function (item) {
+    return item.name === username;
+  });
+
+  return match.length > 0;
+};

